### ISSUE TITLE
incorrect linter errors in version 0.12

### ISSUE THREAD

**Issue by jamiet-msm** [sqlfluff/sqlfluff#3144]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Upgraded to sqlfluff 0.12 and immediately got errors that (a) we didn't get before and (b) I suspect to be invalid.

### Expected Behaviour

SQL file gets linted without error

### Observed Behaviour

sqlfluff throws error

> L:  26 | P:  95 | L025 | Alias 'ceb' is never used in SELECT statement

You'll see form looking at the SQL that I provided below that alias `ceb` is indeed used

```sql
                        select
                                ceb.emailaddresskey,
                                dac.accountkey,
                                row_number() over (partition by dac.currentemailaddresskey order by coalesce(dac.lastprofileupdatedatetime, dac.accountcreationdate) desc) as rnum
                            from `${datalake_project_id}.edw_red.REDDimAccount` as dac
                            inner join `${datalake_project_id}.edw_red.REDDimEmailAddress` as ceb
                                on ceb.emailaddresskey = dac.currentemailaddresskey   --noqa
                            where dac.currentemailaddresskey != -1
                                and coalesce(dac.AccountCreationDate, '1900-01-01') < timestamp_trunc(current_timestamp(), day)
                                and coalesce(dac.LastProfileUpdateDateTime, '1900-01-01') < timestamp_trunc(current_timestamp(), day)
                                and trim(ceb.EmailMarketableStatus) not in ( 'Disposable', 'invalid_account', 'invalid_address', 'invalid_domain', 'possible_spamtrap')
                                and ceb.EmailAddressKey != (-1)
                        
```

therefore I don't think this is valid.

### How to reproduce

write this to a file somewhere:
```sql
with max_key as (
    select max(ExtractAuditKey) as ExtractAuditKey
    from `${project_containing_tables_required_for_braze}.edw.AccountSavedAddressBridge`
),

CRMEmailMyProfileLstAddressKey as (
    select
        abg.emailaddresskey,
        abg.addresskey
    from
        (select
                dac.emailaddresskey,
                abg.addresskey,
                row_number() over (partition by abg.accountkey order by abg.LastUpdatedDateTime desc, abg.addresskey desc) as rnum
            from `${project_containing_tables_required_for_braze}.edw.AccountSavedAddressBridge` as abg
            inner join
                (select     --noqa
                        emailaddresskey,
                        accountkey
                    from
                        (select
                                ceb.emailaddresskey,
                                dac.accountkey,
                                row_number() over (partition by dac.currentemailaddresskey order by coalesce(dac.lastprofileupdatedatetime, dac.accountcreationdate) desc) as rnum
                            from `${datalake_project_id}.edw_red.REDDimAccount` as dac
                            inner join `${datalake_project_id}.edw_red.REDDimEmailAddress` as ceb
                                on ceb.emailaddresskey = dac.currentemailaddresskey   --noqa
                            where dac.currentemailaddresskey != -1
                                and coalesce(dac.AccountCreationDate, '1900-01-01') < timestamp_trunc(current_timestamp(), day)
                                and coalesce(dac.LastProfileUpdateDateTime, '1900-01-01') < timestamp_trunc(current_timestamp(), day)
                                and trim(ceb.EmailMarketableStatus) not in ( 'Disposable', 'invalid_account', 'invalid_address', 'invalid_domain', 'possible_spamtrap')
                                and ceb.EmailAddressKey != (-1)
                        )
                    where rnum = 1) as dac
                on abg.accountkey = dac.accountkey   --noqa
                    and abg.isactive = 'Yes'
                    and coalesce(abg.CreatedDateTime, '1900-01-01') < timestamp_trunc(current_timestamp(), day)
                    and coalesce(abg.LastUpdatedDateTime, '1900-01-01') < timestamp_trunc(current_timestamp(), day)
            where abg.ExtractAuditKey > coalesce((
                        select HWMTimeStamp
                        from `${audit_dataset_id}.${HighWaterMark_table_id}`
                        where Process = 'MSM_IN_MODEL.UserProfile_Events_AttitudinalSegment'
                        order by CompletedTime desc
                        limit 1), 0)
        ) as abg
    where rnum = 1
),

CRMAttitudinalSegment as (
    select
        cma.EmailAddressKey,
        pal.AttitudinalSegment
    from CRMEmailMyProfileLstAddressKey as cma
    inner join `${datalake_project_id}.edw_red.REDDimAddress` as daddr
        on cma.addresskey = daddr.addresskey
    left join `${project_containing_tables_required_for_braze}.edw.PostcodeAttLookup` as pal
              on upper(
                replace(
                    daddr.PostCode,
                    ' ',
                    ''
                )
            ) = upper(
                replace(
                    pal.PostCodeFormat3,
                    ' ',
                    ''
                )
            )
    group by cma.EmailAddressKey, pal.AttitudinalSegment
),

ExtAttitudinalSegment as (
    select
        uc.UserConsentId as ExternalId,
        cas.AttitudinalSegment,
        lower(uc.EmailAddress) as EmailAddress
    from CRMAttitudinalSegment as cas
    inner join `${datalake_project_id}.edw_red.REDFctUserConsent` as uc
        on cas.EmailAddresskey = uc.EmailAddresskey
            and uc.UserConsentUpdatedDateTime >= cast(date_sub(current_date(), interval 3 year) as timestamp)
            and upper(uc.UserConsentBrandName) = 'MONEYSUPERMARKET'
    group by uc.UserConsentId, uc.EmailAddress, cas.AttitudinalSegment
)

select
    a.ExternalId,
    a.EmailAddress,
    a.AttitudinalSegment,
    max(b.ExtractAuditKey) as ExtractAuditKey
from ExtAttitudinalSegment as a
cross join max_key as b
group by a.ExternalId, a.EmailAddress, a.AttitudinalSegment
```

then run 
`sqlfluff lint --dialect bigquery --exclude-rules L016,L031,L014,L057 path/to/file.sql`

### Dialect

bigquery

### Version

0.12

### Configuration

n/a we configure via command line options

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by jamiet-msm** [comment]
We were previously on version 0.11.2 by the way

**Comment by tunetheweb** [comment]
Looks like this was caused by #2993 and we need to enhance rule L025 to take into account that.

**Comment by tunetheweb** [comment]
Simpler reproducible test case:

```sql
select col1
from table1 as abg
inner join (
    select
        ceb.emailaddresskey,
        dac.accountkey
    from table2 as dac
    inner join table3 as ceb
        on ceb.col2 = dac.col2
) as abh
on abg.col1 = abh.col1
```

**Comment by jamiet-msm** [comment]
Looks like I can eliminate the problem by converting to a CTE. From @tunetheweb 's example, this doesn't cause the same error:

```sql
with abh as (
    select
        ceb.emailaddresskey,
        dac.accountkey
    from table2 as dac
    inner join table3 as ceb
        on ceb.col2 = dac.col2
)
select col1
from table1 as abg
inner join  abh
on abg.col1 = abh.col1
```

Nevertheless its still a problem because our upgrade to 0.12 caused many such errors, the example above was just one of the affected files. I'd rather not trawl through all those files and update them.

Thank you for the swift reply @tunetheweb

**Comment by tunetheweb** [comment]
The problem is this code:

https://github.com/sqlfluff/sqlfluff/blob/b8e2b45b47fcf7d2d87222d635123573ee905207/src/sqlfluff/dialects/dialect_ansi.py#L1462-L1463

This used to be:

```py
        # In some dialects, like TSQL, join clauses can have nested join clauses
        for join_clause in self.get_children("join_clause"):
```

So now, while parsing the outer query, it's picking up all the inner joins from sub queries:

So for this parse tree:

```
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    statement:
[L:  1, P:  1]      |        select_statement:
[L:  1, P:  1]      |            select_clause:
[L:  1, P:  1]      |                keyword:                                      'select'
[L:  1, P:  7]      |                [META] indent:
[L:  1, P:  7]      |                whitespace:                                   ' '
[L:  1, P:  8]      |                select_clause_element:
[L:  1, P:  8]      |                    column_reference:
[L:  1, P:  8]      |                        identifier:                           'col1'
[L:  1, P: 12]      |            newline:                                          '\n'
[L:  2, P:  1]      |            [META] dedent:
[L:  2, P:  1]      |            from_clause:
[L:  2, P:  1]      |                keyword:                                      'from'
[L:  2, P:  5]      |                whitespace:                                   ' '
[L:  2, P:  6]      |                from_expression:
[L:  2, P:  6]      |                    [META] indent:
[L:  2, P:  6]      |                    from_expression_element:
[L:  2, P:  6]      |                        table_expression:
[L:  2, P:  6]      |                            table_reference:
[L:  2, P:  6]      |                                identifier:                   'table1'
[L:  2, P: 12]      |                        whitespace:                           ' '
[L:  2, P: 13]      |                        alias_expression:
[L:  2, P: 13]      |                            keyword:                          'as'
[L:  2, P: 15]      |                            whitespace:                       ' '
[L:  2, P: 16]      |                            identifier:                       'abg'
[L:  2, P: 19]      |                    newline:                                  '\n'
[L:  3, P:  1]      |                    [META] dedent:
[L:  3, P:  1]      |                    join_clause:
[L:  3, P:  1]      |                        keyword:                              'inner'
[L:  3, P:  6]      |                        whitespace:                           ' '
[L:  3, P:  7]      |                        keyword:                              'join'
[L:  3, P: 11]      |                        [META] indent:
[L:  3, P: 11]      |                        whitespace:                           ' '
[L:  3, P: 12]      |                        from_expression_element:
[L:  3, P: 12]      |                            table_expression:
[L:  3, P: 12]      |                                bracketed:
[L:  3, P: 12]      |                                    start_bracket:            '('
[L:  3, P: 13]      |                                    [META] indent:
[L:  3, P: 13]      |                                    newline:                  '\n'
[L:  4, P:  1]      |                                    whitespace:               '    '
[L:  4, P:  5]      |                                    select_statement:
[L:  4, P:  5]      |                                        select_clause:
[L:  4, P:  5]      |                                            keyword:          'select'
[L:  4, P: 11]      |                                            [META] indent:
[L:  4, P: 11]      |                                            newline:          '\n'
[L:  5, P:  1]      |                                            whitespace:       '        '
[L:  5, P:  9]      |                                            select_clause_element:
[L:  5, P:  9]      |                                                column_reference:
[L:  5, P:  9]      |                                                    identifier:  'ceb'
[L:  5, P: 12]      |                                                    dot:      '.'
[L:  5, P: 13]      |                                                    identifier:  'emailaddresskey'
[L:  5, P: 28]      |                                            comma:            ','
[L:  5, P: 29]      |                                            newline:          '\n'
[L:  6, P:  1]      |                                            whitespace:       '        '
[L:  6, P:  9]      |                                            select_clause_element:
[L:  6, P:  9]      |                                                column_reference:
[L:  6, P:  9]      |                                                    identifier:  'dac'
[L:  6, P: 12]      |                                                    dot:      '.'
[L:  6, P: 13]      |                                                    identifier:  'accountkey'
[L:  6, P: 23]      |                                        newline:              '\n'
[L:  7, P:  1]      |                                        whitespace:           '    '
[L:  7, P:  5]      |                                        [META] dedent:
[L:  7, P:  5]      |                                        from_clause:
[L:  7, P:  5]      |                                            keyword:          'from'
[L:  7, P:  9]      |                                            whitespace:       ' '
[L:  7, P: 10]      |                                            from_expression:
[L:  7, P: 10]      |                                                [META] indent:
[L:  7, P: 10]      |                                                from_expression_element:
[L:  7, P: 10]      |                                                    table_expression:
[L:  7, P: 10]      |                                                        table_reference:
[L:  7, P: 10]      |                                                            identifier:  'table2'
[L:  7, P: 16]      |                                                    whitespace:  ' '
[L:  7, P: 17]      |                                                    alias_expression:
[L:  7, P: 17]      |                                                        keyword:  'as'
[L:  7, P: 19]      |                                                        whitespace:  ' '
[L:  7, P: 20]      |                                                        identifier:  'dac'
[L:  7, P: 23]      |                                                newline:      '\n'
[L:  8, P:  1]      |                                                whitespace:   '    '
[L:  8, P:  5]      |                                                [META] dedent:
[L:  8, P:  5]      |                                                join_clause:
[L:  8, P:  5]      |                                                    keyword:  'inner'
[L:  8, P: 10]      |                                                    whitespace:  ' '
[L:  8, P: 11]      |                                                    keyword:  'join'
[L:  8, P: 15]      |                                                    [META] indent:
[L:  8, P: 15]      |                                                    whitespace:  ' '
[L:  8, P: 16]      |                                                    from_expression_element:
[L:  8, P: 16]      |                                                        table_expression:
[L:  8, P: 16]      |                                                            table_reference:
[L:  8, P: 16]      |                                                                identifier:  'table3'
[L:  8, P: 22]      |                                                        whitespace:  ' '
[L:  8, P: 23]      |                                                        alias_expression:
[L:  8, P: 23]      |                                                            keyword:  'as'
[L:  8, P: 25]      |                                                            whitespace:  ' '
[L:  8, P: 26]      |                                                            identifier:  'ceb'
```

It's getting this:


```
BARRY analysing:select col1
from table1 as abg
inner join (
    select
        ceb.emailaddresskey,
        dac.accountkey
    from table2 as dac
    inner join table3 as ceb
        on ceb.col2 = dac.col2
) as abh
on abg.col1 = abh.col1:
BARRY table ref:abg:
BARRY table ref:abh:
BARRY checking alias:abg:
BARRY checking alias:abh:
BARRY checking alias:ceb:
BARRY unused alias:ceb:
BARRY analysing:select
        ceb.emailaddresskey,
        dac.accountkey
    from table2 as dac
    inner join table3 as ceb
        on ceb.col2 = dac.col2:
BARRY table ref:dac:
BARRY table ref:ceb:
BARRY checking alias:dac:
BARRY checking alias:ceb:
```

So it found `ceb` as it finds all outer level aliases AND any nested inner join aliases (`cbe` in this case). Note it does not find `dac` which is also in the inner query).

Think I've found a way to stop it doing this. PR coming up...
